### ISSUE TITLE
How does Karmada handle unschedulable Work

### ISSUE THREAD

**Issue by ahmetb** [karmada-io/karmada#1685]
Hi, I am trying to understand what happens if Work is assigned to a worker cluster, but the cluster is not able to run it.

For example, you have a Deployment with `replicas:10` and an anti-affinity set. So the worker cluster gets assigned a chunk of `replicas: 3` but that cluster only has `2` nodes matching the anti-affinity.

As far as I understand, that deployment chunk will always stay at 2/3. 

What prevents Karmada from getting into this situation? As I understand it, Karmada is indifferent to what's in the `Work` object and doesn't do a pass to see if the workload chunk can fit to the cluster (or satisfy scheduling constraints)? 

I appreciate your explanation ahead of time.

**Comment by hanweisen** [comment]
#697 and #521  may solve your problem.

**Comment by RainbowMango** [comment]
Hi @ahmetb Sorry for the late reply, just get back from a public holiday.

Before a workload is assigned to a worker cluster, the `karmada-scheduler` would evaluate how many replicas could be put in based on the available resources and scheduling constraints (such as affinity and anti-affinity) and so on. 
We can say this part is `first layer` schedule.

After the workload is assigned to a worker cluster, if there are any replicas not able to run, we can use the [karmada-descheduler](https://github.com/karmada-io/karmada/blob/master/docs/descheduler.md) to re-balance the unavailable replicas to other clusters. (`karmada-descheduler` unassigns the unavailable replicas and then `karmada-scheduler` will start re-schedule to assign them to other clusters. )

**Comment by duanmengkk** [comment]
> based on the available resources and scheduling constraints (such as affinity and anti-affinity) and so on

Just a remind,It seems the estimator module only support evaluate pod distribution base on the available resources,but not support evaluate pod distribution base on  scheduling constraints (nodeAffinity,podAffinity and so on ).Do you have a plan for it? @RainbowMango

**Comment by duanmengkk** [comment]
> As I understand it, Karmada is indifferent to what's in the `Work` object and doesn't do a pass to see if the workload chunk can fit to the cluster (or satisfy scheduling constraints)?

Karmada have two `ClusterPreferences`:`StaticWeightList `and `DynamicWeight`.In `StaticWeightList ` mode, the number of pods scheduled to the Member cluster is determined based on the user's config in propagationPolicy，while in `DynamicWeight` mode，the karmada-scheduler would evaluate how many replicas could be put in based on the available resources in member cluster(depend on karmada-scheduler-estimator and `--enable-scheduler-estimator=true` in karmada-scheduler‘s command).It seems karmada-scheduler-estimator not support estimatings scheduling constraints (nodeAffinity,podAffinity and so on ) yet, I am not sure. @ahmetb

**Comment by RainbowMango** [comment]
> but not support evaluate pod distribution base on scheduling constraints (nodeAffinity,podAffinity and so on ).Do you have a plan for it?

@Garrybest has a plan.

**Comment by Garrybest** [comment]
Yes, I have some ideas to refactor the estimator. The `anti-affinity` is more complicated than I have thought before. The design is still going on.

**Comment by hanweisen** [comment]
`anti-affinity` is complicated. I think we can add `nodeAffinity` and `nodeSelector` first.

**Comment by ikaven1024** [comment]
Scheduler in `sub-layer` maybe so complex(rather than node podAffinity, podAffinity, and some customer policy) that `karmada-scheduler` cannot cover it. 

I have an idea. After `descheduler` unassigns the unavailable replicas, it will create a crd to record the saturation capacity of clusters.
```yaml
apiVersion: schedule.karmada.io/v1alpha1
Kind: ScheduleConstraint
Name: deployment-nginx-constraint
spec:
  resource:
  - apiVersion: apps/v1
     kind: Deployment
     name: nginx
  constraint:
  - clusterName: cluster1
     maxReplicas: 5
  deadline: 2022-02-22T22:22:22Z
```
And `karmada-scheduler` will execute scheduling work referring to it.

@RainbowMango @Garrybest

**Comment by Garrybest** [comment]
Interesting idea, I think it helps.

**Comment by ikaven1024** [comment]
> Would you like to have a chat in next Karmada Meeting?

Tomorrow？OK, i will attend if available.
